phlib/nativethread.c

```diff
@@ -868,36 +965,172 @@ NTSTATUS PhGetThreadApartmentState(
         oletlsBaseAddress = (PVOID)oletlsDataAddress;
     }
 
-    if (NT_SUCCESS(status) && oletlsBaseAddress)
+    if (!NT_SUCCESS(status))
+        return status;
+
+    if (!oletlsBaseAddress)
     {
-        PVOID apartmentStateOffset;
+        // Return a special error to indicate that we successfully determined
+        // that the thread has no associated COM state. (diversenok)
+        return NTSTATUS_FROM_WIN32(CO_E_NOTINITIALIZED);
+    }
 
-        // Note: Teb->ReservedForOle is the SOleTlsData structure
-        // and ApartmentState is the dwFlags field. (dmex)
+#ifdef _WIN64
+    if (isWow64)
+        apartmentStateOffset = PTR_ADD_OFFSET(oletlsBaseAddress, UFIELD_OFFSET(SOleTlsData32, Flags));
+    else
+        apartmentStateOffset = PTR_ADD_OFFSET(oletlsBaseAddress, UFIELD_OFFSET(SOleTlsData, Flags));
+#else
+    apartmentStateOffset = PTR_ADD_OFFSET(oletlsBaseAddress, UFIELD_OFFSET(SOleTlsData, Flags));
+#endif
+
+    status = NtReadVirtualMemory(
+        ProcessHandle,
+        apartmentStateOffset,
+        ApartmentFlags,
+        sizeof(OLETLSFLAGS),
+        NULL
+        );
+
+    if (!NT_SUCCESS(status))
+        return status;
+
+    if (ComInits)
+    {
+        PVOID comInitsOffset;
```

Comment:
phlib\nativethread.c(1011,1): error C4700: uninitialized local variable 'comInitsOffset' used

Reply:
Oops, sorry, that was an #ifdef'ed typo affecting 32-bit builds. Fixed.